### ISSUE TITLE
CustomOp hangs while converting numpy array to NDArray 

### ISSUE THREAD

**Issue by leopd** [apache/incubator-mxnet#3813]
I'm writing a new `mx.operator.CustomOp` and in the `forward()` method, the engine hangs (no CPU activity) when converting my numpy array to an NDArray: `mxy = mx.nd.array(y)`.

I have MXNET_CPU_WORKER_NTHREADS=2, as the example and docs say, which smells like there's some kind of race condition at play.  And in fact if I turn this off, then the sample code in example/numpy-ops/custom_softmax.py fails in the same operation.

So I'm guessing I'm hitting whatever bug can sometimes be avoided by setting the using multiple worker threads, but in my case that's not a sufficient workaround.  What else can I try?

**Comment by piiswrong** [comment]
`mxy = mx.nd.array(y, ctx=mx.gpu())`?

**Comment by leopd** [comment]
This is all running in CPU, and I'm pretty sure I tried explicitly specifying the CPU context and it didn't help.  

... and of course the current code isn't showing the bug.  I can pull up the older version of the code where it was failing.

**Comment by leopd** [comment]
Code is here: https://github.com/leopd/mxnet/blob/issue-3813/example/issue-3813.py

Reading the [environment variable docs](http://mxnet.io/how_to/env_var.html) I found that setting `MXNET_ENGINE_TYPE` to `NaiveEngine` seems to fix the problem.  And unsurprisingly for a race-condition, it doesn't repro reliably.  Sometimes it fails, sometimes it doesn't.  In fact once I set to NaiveEngine and it starts to work, I haven't seen it fail again on that machine.
